File: credit/credit.c
#include <cs50.h>
#include <stdio.h>

int to_digits(long cn, int dest[256]);
int is_valid(int src[256], int digits);
int is_amex(int src[256], int digits);
int is_master(int src[256], int digits);
int is_visa(int src[256], int digits);

int main(void)
{
    int card_arr[256] = {0};
    long card_number;
    do
    {
        card_number = get_long("Number: ");
    }
    while (card_number < 0);

    int nDigits = to_digits(card_number, card_arr);

    if (!is_valid(card_arr, nDigits))
        printf("INVALID\n");
    else if (is_amex(card_arr, nDigits))
        printf("AMEX\n");
    else if (is_master(card_arr, nDigits))
        printf("MASTERCARD\n");
    else if (is_visa(card_arr, nDigits))
        printf("VISA\n");
    else
        printf("INVALID\n");

    return 0;
}

int to_digits(long cn, int dest[256])
{
    int i = 0;
    while (cn > 0)
    {
        dest[i] = cn % 10;
        cn /= 10;
        i++;
    }
    return i;
}

int is_valid(int src[256], int digits)
{
    int chksm = 0;
    for (int i = 0; i < digits; i++)
    {
        int tmp = src[i] * (1 + (i % 2));
        // tmp % 10 -> 0-9
        // ((tmp / 10) % 10) -> 10s digit, if applicable
        chksm += tmp % 10 + ((tmp / 10) % 10);
    }
    return !(chksm % 10);
}

int is_amex(int src[256], int digits)
{
    return digits == 15 && src[digits - 1] == 3 && (src[digits - 2] == 5 || src[digits - 2] == 7);
}

int is_master(int src[256], int digits)
{
    return digits == 16 && src[digits - 1] == 5 && src[digits - 2] > 0 && src[digits - 2] < 6;
}

int is_visa(int src[256], int digits)
{
    return (digits == 13 || digits == 16) && src[digits - 1] == 4;
}

File: sentimental-credit/credit.py
from cs50 import get_string

def main():
    card = get_string("Number: ").strip()
    if len(card) not in [13, 15, 16] or not luhn(card):
        print("INVALID 1")
    elif is_master(card):
        print("MASTER")
    elif is_visa(card):
        print("VISA")
    elif is_amex(card):
        print("AMEX")
    else:
        print("INVALID 2")

def luhn(card_number):
    sum = []
    for i in range(len(card_number) - 1, -1, -2):
        sum += [int(card_number[i-1]) * 2]
    print(sum)
    sos = 0
    for s in sum:
        for digit in str(s):
            sos += int(digit)
    print(sos)
    sum = 0
    for i in range(len(card_number), 0, -2):
        sum += int(card_number[i-1])
    print(sum378282246310005)
    return (sum + sos) % 10 == 0


def is_amex(card_number):
    if len(card_number) == 15 and card_number[:2] in ["34", "37"]:
        return True
    return False


def is_visa(card_number):
    if len(card_number) in [13, 16] and card_number[0] == "4":
        return True
    return False


def is_master(card_number):
    if len(card_number) == 16 and card_number[:2] in ["51", "52", "53", "54", "55"]:
        return True
    return False

main()
